### DISCUSSION TITLE
v6 How to implement custom validation?

### DISCUSSION THREAD

**Discussion by XIYO-NET** [adonisjs/core#4565] [Help]
Hello, I'm using nodejs for the first time to develop a system and I'm having a problem with it, according to the documentation: Official AdonisJS packages extend VineJS with custom rules. For example, Lucid contributes unique and exists rules to VineJS. exists rules to VineJS.
I tried `username: vine.string().unique({ table: 'users', column: 'username' }) `and it says: checker is not a function. then I looked at the VineJS documentation and it says Creating custom rules, I followed the docs, but I'm having trouble with Extending schema classes, I don't know 
```
VineString.macro('unique', function (this: VineString, options: Options) {
  return this.use(uniqueRule(options))
}) 
```
should be written inside which file? Or where to encapsulate it?

**Comment by gribbl** [comment]
Hello, if you want to use unique and exists rules, it's a bit different from Adonis v5, which was simpler I think. The only thing I found in the documentation is [here](https://docs.adonisjs.com/guides/basics/validation#passing-metadata-to-validators). Am I missing something? 

For example, when creating a user, to make sure the email is unique:

```ts
export const createUserValidator = vine
  .compile(
    vine.object({
      email: vine.string().unique(async (db, value) => {
        const user = await db
          .from('users')
          .where('email', value)
          .first()
        return !user
      })
    })
  )
```

**Reply by XIYO-NET** [reply]
First of all, thank you for your response. But I don't think it's flexible enough to use it directly to operate in the validator. Because I found, based on the VineJS documentation, that extending the VineString class and the unique method would provide a linkable API to call the unique rule. So wanted to try this.
`await request.validateUsing(storeAccountValidator)`
`const schema = vine.object({
  email: vine
    .string()
    .unique({
      table: 'users',
      column: 'email'
    })
})`

**Comment by aikrom** [comment]
You can extend the schema type next to your rule and don't forget to add (extend) the types, otherwise you will get type errors.

**Comment by aikrom** [comment]
About `unique` rule, there is some discussions/issues: [#4356 ](https://github.com/orgs/adonisjs/discussions/4356), [#995](https://github.com/adonisjs/lucid/issues/995) and [PR #1012](https://github.com/adonisjs/lucid/pull/1012)
